### PR TITLE
Test coverage: added sbt plugin to generate scoverage reports

### PR DESCRIPTION
**Pull request by neomatrix369**
Test coverage: added sbt plugin to generate scoverage reports can be invoked via the 'sbt clean jacocoReport test' command.

The above command generates the below info in the console:
```
[info] Lines: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
[info] Instructions: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
[info] Branches: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
[info] Methods: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
[info] Complexity: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
[info] Class: 0% (>= required 0.0%) covered, 0 of 0 missed, OK
```
Will need to dig into how to generate a .csv files with breaddown of package level line coverage - TODO.

Initial steps towards getting the line coverage per package.

All scripts in place to get coverage and save it into a text file inside `target` folder.

Target command to run (example):
```
./get_coverage_for_challenge.sh SUM
```

### CODE REVIEW

File: get_coverage_for_challenge.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+set -e
+set -u
+set -o pipefail
+
+SCRIPT_CURRENT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+CHALLENGE_ID=$1
+JACOCO_TEST_REPORT_XML_FILE="${SCRIPT_CURRENT_DIR}/target/scala-2.12/jacoco/report/jacoco.xml"
+
+sbt --error clean jacoco || true 1>&2
```
[the remark this answers is not in the document]
Comment (neomatrix369) [reply get_coverage_for_challenge.sh]: Done!

File: get_coverage_for_challenge.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#
+# TODO: when one or more tests fails the jacoco coverage xml isn't created so we return 0.
```
[the remark this answers is not in the document]
Comment (neomatrix369) [reply get_coverage_for_challenge.sh]: Done!

File: get_coverage_for_challenge.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#       This is a work-around in case of missing coverage xml file.
+#       Proper solution would be to capture the results of the failing task and do nothing and proceed with execution
+#       Something to pick up at a later stage.
+#
+if [ -f "${JACOCO_TEST_REPORT_XML_FILE}" ]; then
```
[the remark this answers is not in the document]
Comment (neomatrix369) [reply get_coverage_for_challenge.sh]: Done!

File: get_coverage_for_challenge.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    COVERAGE_OUTPUT=$(xmllint --xpath '//package[@name="befaster/solutions/'${CHALLENGE_ID}'"]/counter[@type="INSTRUCTION"]' ${JACOCO_TEST_REPORT_XML_FILE})
+    MISSED=$(echo $COVERAGE_OUTPUT | awk '{print missed, $3}' | tr '="' ' ' | awk '{print $2}')
+    COVERED=$(echo $COVERAGE_OUTPUT | awk '{print missed, $4}' | tr '="' ' '| awk '{print $2}')
+
+    TOTAL_LINES=$((MISSED + $COVERED))
+    echo $(($COVERED * 100 / $TOTAL_LINES))
```
[the remark this answers is not in the document]
Comment (neomatrix369) [reply get_coverage_for_challenge.sh]: Done!

**Comment by neomatrix369** [comment]
All issues raised have been addressed, please do an end-to-end test using the dpnt-coverage scala docker image.

**Comment by julianghionoiu** [comment]
I had fixed an issue where `expr 0` has an exit code of 1, stopping the process.
https://unix.stackexchange.com/questions/63166/bash-e-exits-when-let-or-expr-evaluates-to-0

I have also consolidated the sbt commands into one.

**Pull request merged**